spec/lib/submission_runners/crystal_spec.rb

```diff
@@ -0,0 +1,146 @@
+    let(:account) { contest.accounts.create! }
+
+    let(:problem) do
+      Problem.create_from_files!({
+        contest:     contest,
+        name:        problem_name,
+        input_file:  fixtures.join('ProblemA.in'),
+        output_file: fixtures.join('ProblemA.out'),
+      })
+    end
+
+    let(:submission) {
+      Submission.create_from_file({
+        problem:  problem,
+        account:  account,
+        filename: fixtures.join("#{problem_name}.cr"),
+      }).tap(&:validate!)
+    }
+
+    subject(:runner) { described_class.new submission }
+
+    context 'good entry submission' do
+      let(:problem_name) { 'compiles_and_runs' }
+
+      it 'builds' do
+        b = runner.build
+        expect(b.err).to eq('')
+        expect(b.out).to_not eq(problem.output)
+        expect(b.success?).to be_truthy
+      end
+
+      it 'runs' do
+        runner.build
+        r = runner.run
+        expect(r.err).to eq('')
+        expect(r.out).to eq(problem.output)
+        expect(r.success?).to be_truthy
+      end
+
+      it 'builds and runs with call' do
+        runner.call
+        expect(runner.output).to eq(problem.output)
+        expect(runner.output_type).to eq('success')
+        expect(runner.run_succeeded).to be_truthy
+      end
+    end
+
+    context 'failing entry submission' do
+      let(:problem_name) { 'compiles_and_runs_wrong_answer' }
+
+      it 'builds' do
+        b = runner.build
+        expect(b.err).to eq('')
+        expect(b.out).to_not eq(problem.output)
+        expect(b.success?).to be_truthy
+      end
+
+      it 'runs' do
+        runner.build
+        r = runner.run
+        expect(r.err).to eq('')
+        expect(r.out).to_not eq(problem.output)
+        expect(r.success?).to be_truthy
+      end
+
+      it 'builds and runs with call' do
+        runner.call
+        expect(runner.output).to_not eq(problem.output)
+        expect(runner.output_type).to eq('success')
+        expect(runner.run_succeeded).to be_truthy
+      end
+    end
+
+    context 'compiles and doesn\'t run entry submission' do
+      let(:problem_name) { 'compiles_and_doesnt_run' }
+
+      it 'builds' do
+        b = runner.build
+        expect(b.err).to eq('')
+        expect(b.out).to_not eq(problem.output)
+        expect(b.success?).to be_truthy
+      end
+
+      it 'runs' do
+        runner.build
+        r = runner.run
+        expect(r.err).to_not eq('')
+        expect(r.out).to_not eq(problem.output)
+        expect(r.success?).to be_falsey
+      end
+
+      it 'builds and runs with call' do
+        runner.call
+        expect(runner.output).to_not eq(problem.output)
+        expect(runner.output_type).to eq('run_failure')
+        expect(runner.run_succeeded).to be_falsey
+      end
+    end
+
+    context 'doesn\'t compile entry submission' do
+      let(:problem_name) { 'doesnt_compile' }
+
+      it 'builds' do
+        skip 'Crystal doesn\'t write to STDERR when compilation fails'
```

Comment:
That's actually a big problem. Part of the spec is that we appropriately notify the participant if their code fails to build. Options are (in roughly decreasing ease):
1. Don't support Crystal until their compiler behaves like a regular unix command line tool
2. See if we can redirect stdout to stderr in the compile command
3  Have the build method for the Crystal runner return a custom result object whose `#err` method does the thing we expect
4. Rethink our entire approach to result objects.
I personally vote for 1 or 2.